evm/src/coverage/visitor.rs

```diff
@@ -0,0 +1,397 @@
+use super::{BranchKind, CoverageItem, SourceLocation};
+use ethers::{
+    prelude::sourcemap::SourceMap,
+    solc::artifacts::ast::{self, Ast, Node, NodeType},
+};
+use std::collections::HashMap;
+use tracing::warn;
+
+#[derive(Debug, Default, Clone)]
+pub struct Visitor {
+    /// The source ID containing the AST being walked.
+    source_id: u32,
+    /// The source code that contains the AST being walked.
+    source: String,
+    /// Source maps for this specific source file, keyed by the contract name.
+    source_maps: HashMap<String, SourceMap>,
+
+    /// The contract whose AST we are currently walking
+    context: String,
+    /// The current branch ID
+    // TODO: Does this need to be unique across files?
+    branch_id: usize,
+    /// Stores the last line we put in the items collection to ensure we don't push duplicate lines
+    last_line: usize,
+
+    /// Coverage items
+    items: Vec<CoverageItem>,
+}
+
+impl Visitor {
+    pub fn new(source_id: u32, source: String, source_maps: HashMap<String, SourceMap>) -> Self {
+        Self { source_id, source, source_maps, ..Default::default() }
+    }
+
+    pub fn visit_ast(mut self, ast: Ast) -> eyre::Result<Vec<CoverageItem>> {
+        // Walk AST
+        for node in ast.nodes.into_iter() {
+            if !matches!(node.node_type, NodeType::ContractDefinition) {
+                continue
+            }
+
+            self.visit_contract(node)?;
+        }
+
+        Ok(self.items)
+    }
+
+    fn visit_contract(&mut self, node: Node) -> eyre::Result<()> {
+        let is_contract =
+            node.attribute("contractKind").map_or(false, |kind: String| kind == "contract");
```

Comment:
should we make this more ergonomic on the `Node`, with a `enum Attribute` and `enum Kind` or something?

Reply:
I considered it initially when doing my first AST PR but it was hard to figure out a nice way to do it within the constraints of how it was already designed. I couldn't really see a good way to add this extra data without having a big struct where half of the fields were empty for most nodes without creating node-specific types and moving more towards something like what solang is doing (e.g. a "root" AST node with children that are `Box<something>` and so on all the way down) which might have been to expensive? Interested to hear your thoughts - the visitor here is pretty simple so adjusting it shouldn't be too much work.

Reply:
I see, 
sounds reasonable, we could however think of adding an enum (Known)Attribute that's `AsRef<str>` so we can do `node.attribute(Attribute::Abstract)`

but no real benefit to that I'd assume